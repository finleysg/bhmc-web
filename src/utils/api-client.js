import { queryCache } from "react-query"

const serverURL = process.env.REACT_APP_SERVER_URL

async function client(endpoint, { data, token, headers: customHeaders, ...customConfig } = {}) {
  const config = {
    method: data ? "POST" : "GET",
    body: data ? JSON.stringify(data) : undefined,
    headers: {
      Authorization: token ? `Token ${token}` : undefined,
      "Content-Type": data ? "application/json" : undefined,
      ...customHeaders,
    },
    ...customConfig,
  }

  const parseError = (error) => {
    if (typeof error === "string") {
      return error
    }
    if (error.non_field_errors) {
      return error.non_field_errors[0]
    }
    // TODO
    return error
  }

  return window.fetch(`${serverURL}/${endpoint}`, config).then(async (response) => {
    if (response.status === 401) {
      queryCache.clear()
      // await auth.logout()
      // refresh the page for them
      window.location.assign(window.location)
      return Promise.reject({ message: "Please re-authenticate." })
    }
    if (response.ok) {
      if (response.status !== 204) {
        const data = await response.json()
        return data
      }
      return null
    } else {
      const data = await response.json()
      return Promise.reject(parseError(data))
    }
  })
}

export { client }
