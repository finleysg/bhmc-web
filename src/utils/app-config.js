export const currentEnvironment = +process.env.REACT_APP_CURRENT_ENVIRONMENT
export const stripePublicKey = process.env.REACT_APP_STRIPE_PUBLIC_KEY
export const giphyApiKey = process.env.REACT_APP_GIPHY_API_KEY
export const seniorRateAge = +process.env.REACT_APP_SENIOR_AGE
export const seniorCompetitionAge = +process.env.REACT_APP_SENIOR_COMPETITION_AGE
export const sentryApiKey = process.env.REACT_APP_SENTRY_API_KEY
export const version = process.env.REACT_APP_VERSION

export const maintenanceMode = process.env.REACT_APP_MODE !== "Live"
