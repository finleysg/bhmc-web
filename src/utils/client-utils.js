const API_BASE = process.env.REACT_APP_API_URL
const AUTH_BASE = process.env.REACT_APP_AUTH_URL
const SERVER_BASE = process.env.REACT_APP_SERVER_URL

const normalizeEndpoint = (endpoint) => {
  const [base, querystring] = endpoint.split("?")
  if (!base.startsWith("/") && base.endsWith("/")) {
    return `${base}${querystring ? "?" + querystring : ""}`
  }
  if (!base.startsWith("/") && !base.endsWith("/")) {
    return `${base}/${querystring ? "?" + querystring : ""}`
  }
  if (base.startsWith("/") && base.endsWith("/")) {
    return `${base.substring(1)}${querystring ? "?" + querystring : ""}`
  }
  if (base.startsWith("/") && !base.endsWith("/")) {
    return `${base.substring(1)}/${querystring ? "?" + querystring : ""}`
  }
  throw new Error("Someone messed up their url munging!")
}

const apiUrl = (endpoint) => {
  return `${API_BASE}/${normalizeEndpoint(endpoint)}`
}
const authUrl = (endpoint) => {
  return `${AUTH_BASE}/${normalizeEndpoint(endpoint)}`
}
const serverUrl = (endpoint) => {
  return `${SERVER_BASE}/${normalizeEndpoint(endpoint)}`
}

const parseError = (error) => {
  if (typeof error === "string") {
    return error
  }
  if (error.non_field_errors) {
    return error.non_field_errors[0]
  }
  if (error.detail) {
    return error.detail
  }
  return JSON.stringify(error)
}

export { apiUrl, authUrl, parseError, serverUrl }
