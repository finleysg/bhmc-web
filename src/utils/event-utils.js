import { format, isDate, isValid } from "date-fns"

const getDayName = (dt) => {
  const dow = dt.getDay()
  if (dow === 0) {
    return "Sunday"
  } else if (dow === 1) {
    return "Monday"
  } else if (dow === 1) {
    return "Tuesday"
  } else if (dow === 1) {
    return "Wednesday"
  } else if (dow === 1) {
    return "Thursday"
  } else if (dow === 1) {
    return "Friday"
  } else if (dow === 1) {
    return "Saturday"
  }
}

const getShortDayName = (dt) => {
  const dow = dt.getDay()
  if (dow === 0) {
    return "Sun"
  } else if (dow === 1) {
    return "Mon"
  } else if (dow === 1) {
    return "Tue"
  } else if (dow === 1) {
    return "Wed"
  } else if (dow === 1) {
    return "Thu"
  } else if (dow === 1) {
    return "Fri"
  } else if (dow === 1) {
    return "Sat"
  }
}

const isoDayFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "yyyy-MM-dd")
  }
  return "--"
}

const dayAndDateFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "iii, MMMM do")
  }
  return "--"
}

const dayDateAndTimeFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "iii, MMMM do h:mm aaaa")
  }
  return "--"
}

const dayNameFormat = (dt) => {
  try {
    if (dt && isDate(dt) && isValid(dt)) {
      return format(dt, "iiii")
    }
    console.log(`invalid date: ${dt}`)
    return "--"
  } catch (error) {
    console.log(error)
    return getDayName(dt)
  }
}

const shortDayNameFormat = (dt) => {
  try {
    if (dt && isDate(dt) && isValid(dt)) {
      return format(dt, "iii")
    }
    console.log(`invalid date: ${dt}`)
    return "--"
  } catch (error) {
    console.log(error)
    return getShortDayName(dt)
  }
}

const monthNameFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "MMMM")
  }
  return "--"
}

const shortMonthNameFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "MMM")
  }
  return "--"
}

export {
  dayAndDateFormat,
  dayDateAndTimeFormat,
  dayNameFormat,
  isoDayFormat,
  monthNameFormat,
  shortDayNameFormat,
  shortMonthNameFormat,
}
