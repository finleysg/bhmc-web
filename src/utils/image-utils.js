const serverUrl = process.env.REACT_APP_SERVER_URL

const getPagedPhotosUrl = (tag) => {
  if (tag) {
    return `photos/?page=1&tags=${tag}`
  } else {
    return "photos/?page=1"
  }
}

const imageUrl = (pic) => {
  if (pic.image_url.startsWith("http")) {
    return pic.image_url // production (from Amazon storage)
  }
  return `${serverUrl}${pic.image_url}`
}

const webImageUrl = (pic) => {
  if (pic.web_url.startsWith("http")) {
    return pic.web_url // production (from Amazon storage)
  }
  return `${serverUrl}${pic.web_url}`
}

const mobileImageUrl = (pic) => {
  if (pic.mobile_url.startsWith("http")) {
    return pic.mobile_url // production (from Amazon storage)
  }
  return `${serverUrl}${pic.mobile_url}`
}

export { getPagedPhotosUrl, imageUrl, mobileImageUrl, webImageUrl }
