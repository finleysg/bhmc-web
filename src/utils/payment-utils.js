const transactionFixedCost = 0.3
const transactionPercentage = 0.029

/**
 * Calculates the transaction fee and total from a given subtotal
 * @param {number} subtotal
 */
const calculateFees = (subtotal, excludeTransactionFee = false) => {
  if (subtotal === 0) {
    return {
      subtotal: 0,
      transactionFee: 0,
      total: 0,
    }
  }

  const total = excludeTransactionFee
    ? subtotal
    : (subtotal + transactionFixedCost) / (1.0 - transactionPercentage)
  const transactionFee = excludeTransactionFee ? 0 : total - subtotal

  return {
    subtotal,
    transactionFee,
    total,
  }
}

/**
 * Calculates the amount due from the payment details
 * @param {Payment} payment - The payment record for the event registration flow
 * @param {Map} eventFees - The event fee map from the club event
 */
const getAmountDue = (payment, eventFees, excludeTransactionFee = false) => {
  if (Boolean(payment?.details) && payment.details.length > 0) {
    const subtotal = payment.details
      .map((detail) => eventFees.get(detail.eventFeeId).amount)
      .reduce((f1, f2) => f1 + f2, 0)
    return calculateFees(subtotal, excludeTransactionFee)
  } else {
    return {
      subtotal: 0,
      transactionFee: 0,
      total: 0,
    }
  }
}

/**
 * Calculates the amount due from the payment edits
 * @param {Payment} payment - The payment record for the event registration flow
 * @param {Map} eventFees - The event fee map from the club event
 */
const getAmountChange = (payment, eventFees) => {
  if (Boolean(payment?.edits) && payment.edits.length > 0) {
    const subtotal = payment.edits
      .map((e) => eventFees.get(e.eventFeeId).amount * (e.action === "add" ? 1 : -1))
      .reduce((f1, f2) => f1 + f2, 0)
    return calculateFees(subtotal, true)
  } else {
    return {
      subtotal: 0,
      transactionFee: 0,
      total: 0,
    }
  }
}

/**
 * Create refund objects out of the given reserve slots and fee information.
 * The fee collection includes 0 - many selected fees/payments.
 * @param {ReserveSlot[]} slots - reserve slots with fee collections
 */
const createRefunds = (slots, notes) => {
  const feeDetails = slots.flatMap((slot) => slot.fees)
  return feeDetails
    .filter((fee) => fee.selected)
    .reduce((acc, curr) => {
      const refund = acc.get(curr.payment.id)
      if (refund) {
        refund.refund_amount += curr.eventFee.amount
      } else {
        acc.set(curr.payment.id, {
          payment: curr.payment.id,
          refund_amount: curr.eventFee.amount,
          notes: notes,
        })
      }
      return acc
    }, new Map())
}

export { calculateFees, createRefunds, getAmountChange, getAmountDue }
