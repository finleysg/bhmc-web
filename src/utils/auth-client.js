import { authUrl, parseError } from "./client-utils"

async function client(endpoint, data, token) {
  const config = {
    method: data ? "POST" : "GET",
    body: data ? JSON.stringify(data) : undefined,
    headers: {
      Authorization: token ? `Token ${token}` : undefined,
      "Content-Type": data ? "application/json" : undefined,
    },
  }

  const url = authUrl(endpoint)

  return window.fetch(url, config).then(async (response) => {
    if (response.ok) {
      if (response.status !== 204) {
        const data = await response.json()
        return data
      }
      return null
    } else {
      const data = await response.json()
      return Promise.reject(parseError(data))
    }
  })
}

export { client }
