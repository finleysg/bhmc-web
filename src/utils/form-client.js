import React from "react"

import { useAuth } from "context/auth-context"

import { apiUrl, parseError } from "./client-utils"

async function client(endpoint, formData, token) {
  const config = {
    method: "POST",
    body: formData,
    headers: {
      Authorization: token ? `Token ${token}` : undefined,
    },
  }

  const url = apiUrl(endpoint)

  return window.fetch(url, config).then(async (response) => {
    if (response.ok) {
      if (response.status !== 204) {
        const data = await response.json()
        return data
      }
      return null
    } else {
      const data = await response.json()
      return Promise.reject(parseError(data))
    }
  })
}

function useFormClient() {
  const {
    user: { token },
  } = useAuth()
  if (!token) {
    throw new Error("A form client requires an authenticated user.")
  }
  return React.useCallback((endpoint, formData) => client(endpoint, formData, token), [token])
}

export { useFormClient }
