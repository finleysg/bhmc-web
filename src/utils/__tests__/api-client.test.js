import * as auth from "context/auth-provider"
import { rest, server } from "test/test-server"
import { client } from "utils/api-client"
import { apiUrl } from "utils/client-utils"

jest.mock("react-query")
jest.mock("context/auth-provider")

test("makes GET requests to the given endpoint", async () => {
  const endpoint = "test-endpoint"
  const mockResult = { value: "Trunalimunumaprzure!" }
  server.use(
    rest.get(apiUrl(endpoint), async (req, res, ctx) => {
      return res(ctx.json(mockResult))
    }),
  )
  const result = await client(endpoint)
  expect(result).toEqual(mockResult)
})

test("add auth token when a token is provided", async () => {
  const token = "trunalimunumaprzure"

  let request
  const endpoint = "test-endpoint"
  const mockResult = { value: "hello world" }
  server.use(
    rest.get(apiUrl(endpoint), async (req, res, ctx) => {
      request = req
      return res(ctx.json(mockResult))
    }),
  )
  await client(endpoint, { token })
  expect(request.headers.get("Authorization")).toEqual(`Token ${token}`)
})

test("allows for config overrides", async () => {
  let request
  const endpoint = "test-endpoint"
  const mockResult = { value: "hello world" }
  server.use(
    rest.put(apiUrl(endpoint), async (req, res, ctx) => {
      request = req
      return res(ctx.json(mockResult))
    }),
  )

  const customConfig = {
    method: "PUT",
    headers: { "Content-Type": "fake-type" },
  }
  await client(endpoint, customConfig)
  expect(request.headers.get("Content-Type")).toBe(customConfig.headers["Content-Type"])
})

test("when data is provided, it is stringified and the method defaults to POST", async () => {
  const endpoint = "test-endpoint"
  server.use(
    rest.post(apiUrl(endpoint), async (req, res, ctx) => {
      return res(ctx.json(req.body)) // echo the data passed in
    }),
  )

  const data = {
    foo: "bar",
  }
  const result = await client(endpoint, { data })

  expect(result).toEqual(data)
})

test("automatically logs the user out if the request returns a 401", async () => {
  const endpoint = "test-endpoint"
  const mockResult = { value: "Trunalimunumaprzure!" }
  server.use(
    rest.get(apiUrl(endpoint), async (req, res, ctx) => {
      return res(ctx.status(401), ctx.json(mockResult))
    }),
  )
  const result = await client(endpoint).catch((err) => err)

  expect(result.message).toMatchInlineSnapshot(`undefined`)
  expect(auth.logout).toHaveBeenCalledTimes(1)
})

test("rejects the promise if there is an error", async () => {
  const endpoint = "test-endpoint"
  const mockResult = { non_field_errors: ["Trunalimunumaprzure!"] }
  server.use(
    rest.get(apiUrl(endpoint), async (req, res, ctx) => {
      return res(ctx.status(400), ctx.json(mockResult))
    }),
  )

  await expect(client(endpoint)).rejects.toMatchSnapshot()
})
