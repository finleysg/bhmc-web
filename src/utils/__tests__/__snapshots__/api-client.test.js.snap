// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rejects the promise if there is an error 1`] = `[Error: Trunalimunumaprzure!]`;
